backend/populate_catalog/task.py

```diff
@@ -18,11 +31,79 @@
 CATALOG_KEY = "catalog.json"
 CONTENTS_KEY = "Contents"
 RECORDS_KEY = "Records"
+MESSAGE_ATTRIBUTES_KEY = "messageAttributes"
+
+LOGGER = set_up_logging(__name__)
 
 
 def lambda_handler(event: JsonObject, _context: bytes) -> JsonObject:
     """Main Lambda entry point."""
 
+    LOGGER.debug(dumps({EVENT_KEY: event}))
+
+    for message in event[RECORDS_KEY]:
+        if (
+            message[MESSAGE_ATTRIBUTES_KEY][MESSAGE_ATTRIBUTE_TYPE_KEY][STRING_VALUE_KEY_LOWER]
+            == MESSAGE_ATTRIBUTE_TYPE_ROOT
+        ):
+            handle_root(message[BODY_KEY])
+        elif (
+            message[MESSAGE_ATTRIBUTES_KEY][MESSAGE_ATTRIBUTE_TYPE_KEY][STRING_VALUE_KEY_LOWER]
+            == MESSAGE_ATTRIBUTE_TYPE_DATASET
+        ):
+            handle_dataset(message[BODY_KEY])
+        else:
+            raise UnhandledSQSMessageException("Unhandled SQS message type")
+
+    return {}
+
+
+class UnhandledSQSMessageException(Exception):
+    pass
+
+
+class GeostoreSTACLayoutStrategy(HrefLayoutStrategy):
+    def get_catalog_href(self, cat: Catalog, parent_dir: str, is_root: bool) -> str:
+        original_path = urlparse(cat.get_self_href()).path.rsplit("/", maxsplit=2)
+        if is_root:
+            cat_root = parent_dir
+        else:
+            cat_root = join(parent_dir, original_path[-2])
+
+        return join(cat_root, original_path[-1])
+
+    def get_collection_href(self, col: Collection, parent_dir: str, is_root: bool) -> str:
+        original_path = urlparse(col.get_self_href()).path.rsplit("/", maxsplit=2)
+        assert not is_root
+        col_root = join(parent_dir, original_path[-2])
```

Comment:
You can join more than two file paths, so you could `return join(col_root, *original_path[-2:])`